File: jdbc-template/src/test/java/io/thomastodon/spring/data/examples/house/MySqlHouseRepositoryTest.java
package io.thomastodon.spring.data.examples.house;

import io.thomastodon.spring.data.examples.Utility;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.UUID;

import static java.util.Collections.emptyList;
import static org.assertj.core.api.Assertions.assertThat;

class MySqlHouseRepositoryTest {

    private MySqlHouseRepository repository;

    @BeforeEach
    void setUp() {
        repository = new MySqlHouseRepository(Utility.jdbcTemplate);
    }

    @Test
    @DisplayName("findAll, returns all of the houses")
    void test() {

        House houseOne = new House(UUID.randomUUID(), emptyList());
        House houseTwo = new House(UUID.randomUUID(), emptyList());
        Room roomOne = new Room(UUID.randomUUID(), emptyList());
        Room roomTwo = new Room(UUID.randomUUID(), emptyList());
        Chair chairOne = new Chair(UUID.randomUUID(), emptyList());
        Chair chairTwo = new Chair(UUID.randomUUID(), emptyList());
        Leg legOne = new Leg(UUID.randomUUID());
        Leg legTwo = new Leg(UUID.randomUUID());

        repository.saveHouse(houseOne);
        repository.saveHouse(houseTwo);
        repository.addRoomToHouse(roomOne, houseOne.getUuid());
        repository.addRoomToHouse(roomTwo, houseOne.getUuid());
        repository.addChairToRoom(chairOne, roomOne.getUuid());
        repository.addChairToRoom(chairTwo, roomOne.getUuid());
        repository.addLegToChair(legOne, chairOne.getUuid());
        repository.addLegToChair(legTwo, chairOne.getUuid());

        List<House> houses = repository.findAll();

        assertThat(houses.size()).isEqualTo(2);

        House actualHouseOne = houses
            .stream()
            .filter(house -> house.getUuid().equals(houseOne.getUuid()))
            .findFirst()
            .get();
        assertThat(actualHouseOne.getRooms().size()).isEqualTo(2);

        House actualHouseTwo = houses
            .stream()
            .filter(house -> house.getUuid().equals(houseTwo.getUuid()))
            .findFirst()
            .get();
        assertThat(actualHouseTwo).isEqualTo(new House(houseTwo.getUuid(), emptyList()));

        Room actualRoomOne = actualHouseOne.getRooms()
            .stream()
            .filter(room -> room.getUuid().equals(roomOne.getUuid()))
            .findFirst()
            .get();
        assertThat(actualRoomOne.getChairs().size()).isEqualTo(2);

        Room actualRoomTwo = actualHouseOne.getRooms()
            .stream()
            .filter(room -> room.getUuid().equals(roomTwo.getUuid()))
            .findFirst()
            .get();
        assertThat(actualRoomTwo).isEqualTo(new Room(roomTwo.getUuid(), emptyList()));

        Chair actualChairOne = actualRoomOne.getChairs()
            .stream()
            .filter(chair -> chair.getUuid().equals(chairOne.getUuid()))
            .findFirst()
            .get();
        assertThat(actualChairOne.getLegs().size()).isEqualTo(2);

        Chair actualChairTwo = actualRoomOne.getChairs()
            .stream()
            .filter(chair -> chair.getUuid().equals(chairTwo.getUuid()))
            .findFirst()
            .get();
        assertThat(actualChairTwo).isEqualTo(new Chair(chairTwo.getUuid(), emptyList()));

        assertThat(actualChairOne.getLegs()).containsExactlyInAnyOrder(
            new Leg(legOne.getUuid()),
            new Leg(legTwo.getUuid())
        );
    }

    @AfterEach
    void tearDown() {
        repository.deleteAll();
    }
}

File: jdbc-template/src/test/java/io/thomastodon/spring/data/examples/house/MySqlHouseRepositoryTest.kt
package io.thomastodon.spring.data.examples.house

import io.thomastodon.spring.data.examples.Utility
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.util.UUID

import java.util.Collections.emptyList
import org.assertj.core.api.Assertions.assertThat

internal class MySqlHouseRepositoryTest {

    private lateinit var repository: MySqlHouseRepository

    @BeforeEach
    fun setUp() {
        repository = MySqlHouseRepository(Utility.jdbcTemplate)
    }

    @Test
    @DisplayName("findAll, returns all of the houses")
    fun test() {

        val houseOne = House(UUID.randomUUID(), emptyList())
        val houseTwo = House(UUID.randomUUID(), emptyList())
        val roomOne = Room(UUID.randomUUID(), emptyList())
        val roomTwo = Room(UUID.randomUUID(), emptyList())
        val chairOne = Chair(UUID.randomUUID(), emptyList())
        val chairTwo = Chair(UUID.randomUUID(), emptyList())
        val legOne = Leg(UUID.randomUUID())
        val legTwo = Leg(UUID.randomUUID())

        repository.saveHouse(houseOne)
        repository.saveHouse(houseTwo)
        repository.addRoomToHouse(roomOne, houseOne.uuid)
        repository.addRoomToHouse(roomTwo, houseOne.uuid)
        repository.addChairToRoom(chairOne, roomOne.uuid)
        repository.addChairToRoom(chairTwo, roomOne.uuid)
        repository.addLegToChair(legOne, chairOne.uuid)
        repository.addLegToChair(legTwo, chairOne.uuid)

        val houses = repository.findAll()

        assertThat(houses.size).isEqualTo(2)

        val actualHouseOne = houses.find { house -> house.uuid == houseOne.uuid }!!
        assertThat(actualHouseOne.rooms.size).isEqualTo(2)

        val actualHouseTwo = houses.find { house -> house.uuid == houseTwo.uuid }!!
        assertThat(actualHouseTwo).isEqualTo(House(houseTwo.uuid, emptyList()))

        val actualRoomOne = actualHouseOne.rooms.find { room -> room.uuid == roomOne.uuid }!!
        assertThat(actualRoomOne.chairs.size).isEqualTo(2)

        val actualRoomTwo = actualHouseOne.rooms.find { room -> room.uuid == roomTwo.uuid }!!
        assertThat(actualRoomTwo).isEqualTo(Room(roomTwo.uuid, emptyList()))

        val actualChairOne = actualRoomOne.chairs.find { chair -> chair.uuid == chairOne.uuid }!!
        assertThat(actualChairOne.legs.size).isEqualTo(2)

        val actualChairTwo = actualRoomOne.chairs.find { chair -> chair.uuid == chairTwo.uuid }!!
        assertThat(actualChairTwo).isEqualTo(Chair(chairTwo.uuid, emptyList()))

        assertThat(actualChairOne.legs).containsExactlyInAnyOrder(
            Leg(legOne.uuid),
            Leg(legTwo.uuid)
        )
    }

    @AfterEach
    fun tearDown() {
        repository.deleteAll()
    }
}